File: dsa-practice-ram/dsa-practice-ram/Arrays/MissingNumbers.cs
namespace DsaPractice.Arrays
{
    internal class MissingNumbers
    {
        public void FindMissingNumbers()
        {
            int[] input = { 4, 3, 2, 7, 8, 2, 3, 1, 10, 13 };

            Array.Sort(input);

            // 1, 2, 2, 3, 3, 4, 7, 8
            List<int> result = new List<int>();

            for (int i = 1; i < input.Length; i++)
            {
                int curr = input[i];
                int previous = input[i - 1];

                for (int j = previous + 1; j < curr; j++)
                {
                    result.Add(j);
                }
            }
            Console.WriteLine(string.Join(",  ", result));
        }

        public void FindMissingNumbersOptimalInplace()
        {
            // To find missing elements between 1 to n, the array should have n number of elements
            // Also the array range should be between 1 to n
            // Does not work if the array have negative numbers
            // Floating point numbers or strings

            int[] input = { 4, 3, 2, 7, 8, 2, 3, 1, 9, 9, 9 };

            for (int i = 0; i < input.Length; i++)
            {
                int index = Math.Abs(input[i]) - 1;

                if (input[index] > 0)
                    input[index] = -input[index];
            }

            var missingNumbers = new List<int>();
            for (int i = 0; i < input.Length; i++)
            {
                if (input[i] > 0)
                    missingNumbers.Add(i + 1);
            }

            Console.WriteLine(string.Join(", ", missingNumbers));
        }
    }
}

File: MyLearning/PythonPractice/dsa_practice/arrays/missing_numbers.py
"""Missing numbers utilities."""

# C# MissingNumbers.FindMissingNumbers

def find_missing_numbers(input_list=None):
    """Return missing integers between consecutive sorted elements.

    Problem: Given an array, find numbers missing between sorted neighbors.
    Approach: Sort the array and fill gaps between adjacent values.
    Time complexity: O(n log n)
    Space complexity: O(n) for results
    """
    if input_list is None:
        input_list = [4, 3, 2, 7, 8, 2, 3, 1, 10, 13]

    if not input_list:
        return []

    nums = list(input_list)
    nums.sort()

    result = []
    for i in range(1, len(nums)):
        curr = nums[i]
        previous = nums[i - 1]
        for value in range(previous + 1, curr):
            result.append(value)

    return result


# C# MissingNumbers.FindMissingNumbersOptimalInplace

def find_missing_numbers_optimal_inplace(input_list=None):
    """Return missing numbers in the range 1..n using in-place marking.

    Problem: For an array of size n with values in 1..n, find missing values.
    Approach: Mark indices by negating values, then collect positives.
    Time complexity: O(n)
    Space complexity: O(1) extra (in-place on input list)
    """
    if input_list is None:
        input_list = [4, 3, 2, 7, 8, 2, 3, 1, 9, 9, 9]

    nums = input_list
    for i in range(len(nums)):
        index = abs(nums[i]) - 1
        if nums[index] > 0:
            nums[index] = -nums[index]

    missing_numbers = []
    for i in range(len(nums)):
        if nums[i] > 0:
            missing_numbers.append(i + 1)

    return missing_numbers
